### PR TITLE
8287843: File::getCanonicalFile doesn't work for \\?\C:\ style paths DOS device paths

### PR DESCRIPTION
**Pull request by bplb**
In the Windows implementation of java.io.File.getCanonicalPath, strip any long path or UNC prefix before canonicalizing the remainder of the pathname.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8317555](https://bugs.openjdk.org/browse/JDK-8317555) to be approved

### Issues
 * [JDK-8287843](https://bugs.openjdk.org/browse/JDK-8287843): File::getCanonicalFile doesn't work for \\?\C:\ style paths DOS device paths (**Bug** - P4)
 * [JDK-8317555](https://bugs.openjdk.org/browse/JDK-8317555): File::getCanonicalFile doesn't work for \\?\C:\ style paths DOS device paths (**CSR**)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15603/head:pull/15603` \
`$ git checkout pull/15603`

Update a local copy of the PR: \
`$ git checkout pull/15603` \
`$ git pull https://git.openjdk.org/jdk.git pull/15603/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15603`

View PR using the GUI difftool: \
`$ git pr show -t 15603`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15603.diff">https://git.openjdk.org/jdk/pull/15603.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15603#issuecomment-1709171610)

### CODE REVIEW

**Comment by bplb** [comment]
> No `File` methods aside from `getCanonicalPath` are affected.

That is to say, other than `GetCanonicalFile`, of course.

**Comment by AlanBateman** [comment]
Would it be possible to summarise how this behaves when File just presents this prefix? It would be a weird thing to do but it may come about from usages of getParent or other methods. I think I'm mostly asking if is possible for the method to return the canonical path of the current directory when it should fail.

**Comment by bplb** [comment]
> Would it be possible to summarise how this behaves when File just represents a prefix?

Baseline:
```
jshell> File f = new File("\\\\?\\")
f ==> \\?

jshell> f.getCanonicalPath()
|  Exception java.io.IOException: Bad pathname
|        at WinNTFileSystem.canonicalize0 (Native Method)
|        at WinNTFileSystem.canonicalize (WinNTFileSystem.java:463)
|        at File.getCanonicalPath (File.java:626)
|        at (#2:1)

jshell> f = new File("\\\\?\\UNC\\")
f ==> \\?\UNC

jshell> f.getCanonicalPath()
|  Exception java.io.IOException: Bad pathname
|        at WinNTFileSystem.canonicalize0 (Native Method)
|        at WinNTFileSystem.canonicalize (WinNTFileSystem.java:463)
|        at File.getCanonicalPath (File.java:626)
|        at (#4:1)
```

Proposed patch:
```
jshell> File f = new File("\\\\?\\")
f ==> \\?

jshell> f.getCanonicalPath()
|  Exception java.io.IOException: Bad pathname
|        at WinNTFileSystem.canonicalize0 (Native Method)
|        at WinNTFileSystem.canonicalize (WinNTFileSystem.java:472)
|        at File.getCanonicalPath (File.java:626)
|        at (#2:1)

jshell> f = new File("\\\\?\\UNC\\")
f ==> \\?\UNC

jshell> f.getCanonicalPath()
$4 ==> "C:\\Users\\bpb\\dev\\jdk\\jdk\\UNC"
```
Presumably the constructor removes the trailing `\\`, hence I suppose the special case of `\\?\UNC` should be checked in the code right after the proposed change.

**Comment by AlanBateman** [comment]
The two cases that I'm wondering about are `\\?\` and `\\?\UNC`.

For `\\?\` it looks like it will throw (you've got a test for that. I'm just wondering if there is input today where getParent().getCanonicalPath() would succeed today and fail wit hate change.

For `\\?\UNC` it looks it will be converted by the new code to `\` so the root directory of the current volume, do I have that right? I'm trying to see how this case fails in the badPaths list.

**Comment by bplb** [comment]
> I'm trying to see how this case fails in the badPaths list.

I'll investigate the other topics you raise, but I should point out that the results of this test are the same as for the mainline except for the last two good paths
```
        list.add(Arguments.of("\\\\?\\C:\\Users\\file.dat",
                              "C:\\Users\\file.dat"));
        list.add(Arguments.of("\\\\?\\UNC\\localhost\\C$\\Users\\file.dat",
                              "\\\\localhost\\C$\\Users\\file.dat"));
```
which fail with the mainline.

**Comment by bplb** [comment]
> The two cases that I'm wondering about are `\\\\?\\` and `\\\\?\\UNC`.

Both of these cases throw an IOE with message "Bad pathname" for `getCanonicalPath()` and `getParentFile().getCanonicalPath()` with both the mainline and with this change.

**Comment by AlanBateman** [comment]
For `\\\\?\\`, the normalization will drop the trailing slash so the path string is actually `\\\\?`. I that that case is okay.

For input such as `\\\\?\\foo` and `\\\\?\\C:`, the stripping results in a relative path, `foo` and `C:` in these examples, which means the canonicalize code is starting with a relative path. We have to be careful here as the path has always been make absolute before attempting to canonicalize.

For `\\\\?\\UNC` and `\\\\?\\UNC\\`, the stripping means these will be treated as `\\`, which is the root directory of the current volume. I know these will fail with "The network path was not path" but I can't help thinking it would reject/throw early.

**Comment by bplb** [comment]
> These also fail with "Bad pathname" with the master but give the erroneous results `C:\\<user.dir>\\foo` and `C:`, respectively, with the patch. This needs to be fixed.

The problem here might be that `File.getCanonicalPath` resolves the path
```
return FS.canonicalize(FS.resolve(this));
```
before calling `FS.canonicalize`. It could be that if after stripping the `\\\\?\\` prefix the result is relative or drive-relative, then it should first be resolved in the same way before proceeding.

**Comment by AlanBateman** [comment]
> It could be that if after stripping the `\\\\?\\` prefix the result is relative or drive-relative, then it should first be resolved in the same way before proceeding.

Right, which is why the one-arg WinNTFileSystem.resolve may be the right place to check for the prefix as it would fix the issue for both getAbsolutePath too.

**Comment by bplb** [comment]
> the one-arg WinNTFileSystem.resolve may be the right place to check for the prefix

So changed in a8726fbb8a070388f2b9756ee88c746b12c18397 which also adds a couple of test cases. Perhaps some cases should be added to the `GetAbsolutePath` test as well.

**Comment by AlanBateman** [comment]
> So changed in [a8726fb](https://github.com/openjdk/jdk/commit/a8726fbb8a070388f2b9756ee88c746b12c18397) which also adds a couple of test cases. Perhaps some cases should be added to the `GetAbsolutePath` test as well.

Yes, I think it will need tests. The next thing is ask is the behavior of File::isAbsolute, it looks like it will return true for input like `\\\\?\\foo` but it will be treated by toAbsolutePath as a relative path.

**Comment by bplb** [comment]
> File::isAbsolute, it looks like it will return true for input like `\\\\?\\foo` but it will be treated by toAbsolutePath as a relative path.

Right on:
```
jshell> File f = new File("\\\\?\\foo")
f ==> \\?\foo

jshell> f.isAbsolute()
$2 ==> true

jshell> f.getAbsolutePath()
$3 ==> "C:\\Users\\bpb\\foo"

jshell> f.getCanonicalPath()
$4 ==> "C:\\Users\\bpb\\foo"
```

**Comment by AlanBateman** [comment]
> > File::isAbsolute, it looks like it will return true for input like `\\\\?\\foo` but it will be treated by toAbsolutePath as a relative path.
> 
> Right on:

Ideally the prefix would be removed at the front door, meaning when creating the File object, but it may be 20 years too late to do that change. From a compatibility perspective then limiting the it to canonicalization should be okay but it the path needs to make absolute again after stripping.

**Comment by bplb** [comment]
It might be that the long prefix needs to be stripped local to several different methods. Will investigate.

**Comment by bplb** [comment]
Support for `isAbsolute` was added. All `jdk_core` tests still pass. Test cases still need to be added to `GetAbsolutePath.java` and `IsAbsolute.java`. These tests also appear ripe for conversion to JUnit 5.

**Comment by AlanBateman** [comment]
> Support for `isAbsolute` was added. 

Okay, this looks better but I'm still concerned about other potential inconsistencies. Can you try out getParent with different input to see if any issues come up related to the prefix length.

**Comment by AlanBateman** [comment]
@bplb and I discussed this issue yesterday and agreed that the right thing is to strip the prefix at the front door (meaning the normalization of the initial path string), all other directions appear to end up with inconsistencies and surprising behavior after several hops. Doing the right thing here means a behavior change for environments where the path string comes from native code using the long path or UNC prefix. This will need a CSR and release note. We agreed to reevaluate the compatibility impact before JDK 22 RDP1.

**Comment by AlanBateman** [comment]
> remove stripLongOrUNCPrefix() except in normalize()

I haven't had time to look at the test changes yet but the I think the code changes look okay so I think go ahead and create the CSR as this behavior change will need to be tracked.

**Comment by AlanBateman** [comment]
> The CSR was created.

Thanks. I changed the "Compatibility Risk" to medium and expanded its description to more clearly explain the compatibility risk to fix this long standing issue.

File: test/jdk/java/io/File/WindowsPrefixes.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8287843
+ * @summary Basic test for Windows path prefixes
+ * @requires (os.family == "windows")
```
Comment (AlanBateman) [inline test/jdk/java/io/File/WindowsPrefixes.java:27]: Given the `@requires` then I assume `@EnabledOnOs(OS.WINDOWS)` can be dropped from the test methods.

File: test/jdk/java/io/File/WindowsPrefixes.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    private static Stream<Arguments> paths() {
+        return Stream.of(Arguments.of(""),
+                         Arguments.of("C:\\"),
+                         Arguments.of("C:"),
+                         Arguments.of("foo"),
+                         Arguments.of("foo\\bar"),
+                         Arguments.of("C:\\foo"),
+                         Arguments.of("C:foo"),
+                         Arguments.of("C:\\foo\\bar"));
```
Comment (AlanBateman) [inline test/jdk/java/io/File/WindowsPrefixes.java:54]: I suppose this should include `\foo` in the list too.

File: test/jdk/java/io/File/GetAbsolutePath.java [hunk 3]
```diff
@@ -22,74 +22,89 @@
...
+                         Arguments.of("\\\\?\\" + drive + ":", USER_DIR),
+                         Arguments.of("\\\\?\\" + drive + ":bar", USER_DIR + "\\bar"));
+    }
+
+    @EnabledOnOs(OS.WINDOWS)
+    @ParameterizedTest
+    @MethodSource("windowsSource")
+    public void windows(String path, String absolute) throws IOException {
+        assertEquals(absolute.toLowerCase(),
+                     new File(path).getAbsolutePath().toLowerCase());
```
Comment (AlanBateman) [inline test/jdk/java/io/File/GetAbsolutePath.java:76]: WinNTFileSystem.compareTo uses compareToIgnoreCase so this test should probably do the same.

**Review by AlanBateman — COMMENTED** [review]
Implementation change looks fine. I see CSR is approved. We'll need to RN this change too.

File: test/jdk/java/io/File/WindowsPrefixes.java (hunk 1 above)
Comment (bplb) [reply test/jdk/java/io/File/WindowsPrefixes.java:27]: Indeed the `@EnabledOnOs` are pointless here. Removed in 2f31318530d7093dc69c07115c5d46b63612ff4b.

File: test/jdk/java/io/File/WindowsPrefixes.java (hunk 2 above)
Comment (bplb) [reply test/jdk/java/io/File/WindowsPrefixes.java:55]: Included in 2f31318530d7093dc69c07115c5d46b63612ff4b.

File: test/jdk/java/io/File/GetAbsolutePath.java (hunk 3 above)
Comment (bplb) [reply test/jdk/java/io/File/GetAbsolutePath.java]: So changed in 2f31318530d7093dc69c07115c5d46b63612ff4b.

**Review by AlanBateman — APPROVED** [review]
I don't have any other comments on this. I assume you'll run the tests a number of times to make sure they are stable.

**Comment by bplb** [comment]
> I assume you'll run the tests a number of times to make sure they are stable.

I always run tiers 1-3 before integrating, but will add also separate runs of these tests with repeats.

**Comment by bplb** [comment]
Tiers 1-3 tests passed on Windows. The tests in this PR passed 150 repeats and the other jdk_io tests 50 repeats on Windows. The tests which are not Windows-specific were verified to still succeed on Linux.

**Pull request closed without merging**